optimize_ast.ts

```diff
@@ -0,0 +1,343 @@
+                return {tag: "literal", value, a: expr.a};
+            }
+            return {...expr, left: optLeft, right: optRight};
+        case "method-call":
+            var optArgs = expr.arguments.map(e => optimizeExpr(e));
+            return {...expr, arguments: optArgs};
+        default:
+            return {...expr};
+    }
+}
+
+function optimizeStmt(stmt: Stmt<[Type, SourceLocation]>): Stmt<[Type, SourceLocation]> {
+    switch (stmt.tag){
+        case "assign":
+            var optValue = optimizeExpr(stmt.value);
+            return {...stmt, value: optValue};
+        case "expr":
+            var optExpr = optimizeExpr(stmt.expr);
+            return {...stmt, expr: optExpr};
+        case "if":
+            return {...stmt};
+        case "return":
+            var optValue = optimizeExpr(stmt.value);
+            return {...stmt, value: optValue};
+        case "while":
+            var optCond = optimizeExpr(stmt.cond);
+            var optBody = stmt.body.map(stmt => optimizeStmt(stmt));
+            return {...stmt, cond: optCond, body: optBody};
+        case "field-assign":
+            var optValue = optimizeExpr(stmt.value);
+            return {...stmt, value: optValue};
+        default:
+            return {...stmt};
+    }
+}
+
+function foldBuiltin2(lsh: Literal, rhs: Literal, name: string): Literal {
+    switch (name) {
+        case "max":
+            if (lsh.tag === "num" && rhs.tag === "num")
+                return {tag: "num", value: Math.max(lsh.value, rhs.value)};
+            return {tag: "none"};
+        case "min":
+            if (lsh.tag === "num" && rhs.tag === "num")
+                return {tag: "num", value: Math.min(lsh.value, rhs.value)};
+            return {tag: "none"};
+        case "pow":
+            if (lsh.tag === "num" && rhs.tag === "num")
+                return {tag: "num", value: Math.pow(lsh.value, rhs.value)};
+            return {tag: "none"};
+        default:
+            return {tag: "none"};
+    }
+}
+
+function foldBinop(lhs: Literal, rhs: Literal, op: BinOp): Literal{
+    switch(op) {
+        case BinOp.Plus:
+            if(lhs.tag !== "num" || rhs.tag !== "num"){
+                return {tag: "none"};
+            }  
+            return {tag: "num", value: lhs.value + rhs.value};
+        case BinOp.Minus:
+            if(lhs.tag !== "num" || rhs.tag !== "num"){
+                return {tag: "none"};
+            }  
+            return {tag: "num", value: lhs.value - rhs.value};
+        case BinOp.Mul:
+            if(lhs.tag !== "num" || rhs.tag !== "num"){
+                return {tag: "none"};
+            }  
+            return {tag: "num", value: lhs.value * rhs.value};
+        case BinOp.IDiv:
+            if(lhs.tag !== "num" || rhs.tag !== "num"){
+                return {tag: "none"};
+            }  
+            return {tag: "num", value: lhs.value / rhs.value};
```

Comment:
so when you do `lhs.value / rhs.value` in JS, does it exactly mimic `IDiv`?  for example what would be `4/3` after your binary fold, and what would its IDiv result? aren't you probably sending back floats? wouldn't that cause an issue? in general think of the type of data you return from your JS constant folding functions.

Reply:
Thanks for pointing it out, that should be a mistake. We will fix that and add more comprehensive tests to cover all cases (hopefully).